featurebyte/routes/user_defined_function/controller.py

```diff
@@ -0,0 +1,226 @@
+        """
+        # validate feature store id exists
+        feature_store = await self.feature_store_service.get_document(
+            document_id=data.feature_store_id
+        )
+
+        # create user defined function
+        user_defined_function = await self.service.create_document(data)
+        try:
+            # check if user defined function exists in warehouse
+            await self.feature_store_warehouse_service.check_user_defined_function_exists(
+                user_defined_function=user_defined_function,
+                feature_store=feature_store,
+                get_credential=get_credential,
+            )
+        except Exception as exc:
+            # if not exists, delete the user defined function
+            await self.service.delete_document(document_id=user_defined_function.id)
+            raise DocumentCreationError(f"User defined function not exists: {exc}") from exc
+
+        return user_defined_function
+
+    async def update_user_defined_function(
+        self,
+        document_id: PydanticObjectId,
+        data: UserDefinedFunctionUpdate,
+    ) -> UserDefinedFunctionModel:
+        """
+        Update UserDefinedFunction at persistent
+
+        Parameters
+        ----------
+        document_id: PydanticObjectId
+            UserDefinedFunction id
+        data: UserDefinedFunctionUpdate
+            UserDefinedFunction update payload
+
+        Returns
+        -------
+        UserDefinedFunctionModel
+            Updated user_defined_function object
+
+        Raises
+        ------
+        DocumentUpdateError
+            If user defined function used in any saved feature
+        """
+        # check if user defined function exists
+        document = await self.service.get_document(document_id=document_id)
+
+        # check if function used in any saved feature
+        features = await self.feature_service.list_documents(
+            query_filter={"user_defined_function_ids": {"$in": [document_id]}},
+        )
+        if features["total"]:
+            feature_names = [doc["name"] for doc in features["data"]]
+            raise DocumentUpdateError(
+                f"User defined function used by saved feature(s): {feature_names}"
+            )
+
+        func_param_mapping = OrderedDict()
+        for func_param in document.function_parameters:
+            func_param_mapping[func_param.name] = func_param
+
+        for func_param in data.function_parameters:
+            if func_param.name in func_param_mapping:
+                func_param_mapping[func_param.name] = func_param
+            else:
+                raise DocumentUpdateError(f"Function parameter not exists: {func_param.name}")
+
+        function_parameters = list(func_param_mapping.values())
+        if function_parameters == document.function_parameters:
+            raise DocumentUpdateError("No changes found in function parameters")
+
+        updated_document = await self.service.update_document(
```

Comment:
Do we have to validate that the new parameters are actually valid for the function defined in the data warehouse?

Reply:
Yes, we should validate it. Good catch.